equational_theories/EquationalResult.lean

```diff
@@ -87,35 +92,61 @@ def extractEquationalResults {m : Type → Type} [Monad m] [MonadEnv m] [MonadEr
     m (Array Entry) := do
   return equationalResultsExtension.getState (← getEnv)
 
+def extractTheorems {m : Type → Type} [Monad m] [MonadEnv m] [MonadError m] :
+    m (Array Entry) := do
+  return (equationalResultsExtension.getState (← getEnv)).filter (·.proven)
+
+def extractConjectures {m : Type → Type} [Monad m] [MonadEnv m] [MonadError m] :
+    m (Array Entry) := do
+  return (equationalResultsExtension.getState (← getEnv)).filter (!·.proven)
+
 /-- Prints the contents of the equational results environment extension.
 -/
 syntax (name := printEquationalResults) "#print_equational_results" : command
 
 elab_rules : command
 | `(command| #print_equational_results) => do
-  let rs ← extractEquationalResults
-  for ⟨name, _filename, res⟩ in rs do
+  let rs ← extractTheorems
+  for ⟨name, _filename, res, _⟩ in rs do
     match res with
     | .implication ⟨lhs, rhs⟩ => println! "{name}: {lhs} → {rhs}"
     | .facts ⟨satisfied, refuted⟩ => println! "{name}: {satisfied} // {refuted}"
     | .unconditional rhs => println! "{name}: {rhs} holds unconditionally"
 
---- Output of the extract_implications executable.
-structure Output where
-  implications : Array Implication
-  facts : Array Facts
-  unconditionals : Array String
-deriving Lean.ToJson, Lean.FromJson
-
-def collectResults {m : Type → Type} [Monad m] [MonadEnv m] [MonadError m] :
-    m Output := do
-  let rs := equationalResultsExtension.getState (← getEnv)
-  let mut implications : Array Implication := #[]
-  let mut facts : Array Facts := #[]
-  let mut unconditionals : Array String := #[]
-  for ⟨_name, _filename, res⟩ in rs do
-    match res with
-    | .implication imp => implications := implications.push imp
-    | .facts fact => facts := facts.push fact
-    | .unconditional s => unconditionals := unconditionals.push s
-  return ⟨implications, facts, unconditionals⟩
+end Result
+
+namespace Conjecture
+
+/--
+Like `proof_wanted` from Batteries, but records the conjecture in an
+environment extension.
+-/
+@[command_parser]
+def «conjecture» := leading_parser
+  declModifiers false >> "conjecture" >> declId >> ppIndent declSig
+
+/-- Elaborates a `conjecture` declaration. The declaration is translated to an axiom during
+elaboration, but it's then removed from the environment.
+-/
+@[command_elab «conjecture»]
+def elabConjecture : CommandElab
+  | `($mods:declModifiers conjecture $name $args* : $res) => do
+    let maybe_entry ← withoutModifyingEnv do
+      let original_length := (equationalResultsExtension.getState (← getEnv)).size
+
+      -- The helper axiom is used instead of `sorry` to avoid spurious warnings
+      elabDeclaration <| ← `(command| axiom helper (p : Prop) : p)
+      elabDeclaration <| ← `(command| $mods:declModifiers
+                                      theorem $name $args* : $res := helper _)
+
+      if original_length < (equationalResultsExtension.getState (← getEnv)).size then
```

Comment:
Please add a comment explaining this. The `equational_results` extension will have been updated if the conjecture was tagged with the attribute, and we want to pull out the entry so we can durably add it to the extension in the outer environment.